src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -303,34 +303,24 @@ bool TryGetFileMappingFromMetadataImporter(FileKey fileKey, [NotNullWhen(true)]
         FileKey fileKey,
         Func<FileKey, (ModuleMetadata moduleMetadata, TemporaryStorageStreamHandle? storageHandle)> moduleMetadataFactory)
     {
-        var (manifestModule, manifestHandle) = moduleMetadataFactory(fileKey);
-
         using var _1 = ArrayBuilder<ModuleMetadata>.GetInstance(out var moduleBuilder);
         using var _2 = ArrayBuilder<TemporaryStorageStreamHandle?>.GetInstance(out var storageHandles);
 
-        string? assemblyDir = null;
+        var (manifestModule, manifestHandle) = moduleMetadataFactory(fileKey);
+        moduleBuilder.Add(manifestModule);
+        storageHandles.Add(manifestHandle);
+
+        var assemblyDir = Path.GetDirectoryName(fileKey.FullPath);
         foreach (var moduleName in manifestModule.GetModuleNames())
         {
-            if (assemblyDir is null)
-            {
-                moduleBuilder.Add(manifestModule);
-                assemblyDir = Path.GetDirectoryName(fileKey.FullPath);
```

Comment:
3rd: the core place i originally missed was here.  Specifically, we would go in here and add the primary module to the modulebuilder, but totally lose the handle.

--

However, this code is entirely overwrought.  The logic was:

1. Do not add the primary items to the collections.
2. if we have sub modules, add the primary items to the collections, then add hte sub modules.
3. if we had no sub modules, add the primary items to the collections.

so the entire logic was convoluted, and always wanted to add the primary items as the first items.  So that's just what the new logic does.

Reply:
the new logic is trivial and sane.  We get the info abotu the primary module and add to the builders. Then we ge tthe info about the sub-modules, and add to hte builders.  That's it :)